src/proper_sa.erl

```diff
@@ -28,80 +28,178 @@
-         last_generated    = null :: proper_gen:instance()
-        }).
+         last_generated    = null :: proper_gen:instance()}).
+-type sa_target() :: #sa_target{}.
 
 -record(sa_data,
-        {target = undefined                          :: proper_target:target() | 'undefined',
-         %% max runs
-         k_max = 0                                   :: k(),
-         %% run number
-         k_current = 0                               :: k(),
-         %% acceptance probability
-         p = fun (_, _, _) -> false end              :: accept_fun(),
-         %% energy level
-         last_energy = null                          :: proper_target:fitness() | null,
-         last_update = 0                             :: integer(),
-         %% temperature function
-         temperature = 1.0                           :: proper_gen_next:temperature(),
-         temp_func = fun(_, _, _, _, _) -> 1.0 end   :: temp_fun(),
-         %% output function
-         output_fun = fun (_, _) -> ok end            :: output_fun()}).
+        {%% search steps
+         k_max = 0                                 :: k(),
+         %% current step
+         k_current = 0                             :: k(),
+         %% acceptance function
+         p = fun (_, _, _) -> false end            :: accept_fun(),
+         %% fitness
+         last_energy = null                        :: proper_target:fitness() | null,
+         last_update = 0                           :: integer(),
+         %% temperature
+         temperature = 1.0                         :: proper_gen_next:temperature(),
+         temp_func = fun(_, _, _, _, _) -> 1.0 end :: temp_fun()}).
+-type sa_data() :: #sa_data{}.
+
+%% -----------------------------------------------------------------------------
+%% proper_target callbacks
+%% -----------------------------------------------------------------------------
+
+%% Initialize the strategy data based on the 
+%% number of the search steps and the strategy.
+
+%% @private
+-spec init_strategy(proper:setup_opts()) -> sa_data().
+init_strategy(#{numtests := Steps}) ->
+  #sa_data{k_max = Steps,
+           p = get_acceptance_function(),
+           temp_func = get_temperature_function()}.
+
+%% Initialize target state based on the initial generator
+%% and the neighbourhood function.
+
+%% @private
+-spec init_target(proper_target:tmap()) -> sa_target().
+init_target(#{first := First, next := Next}) ->
+  {ok, InitialValue} = proper_gen:safe_generate(First),
+  #sa_target{first = First, next = Next, last_generated = InitialValue}.
+
+%% The function which generates the next instances of
+%% the targeted generator. It also updates the target state.
+
+%% @private
+-spec next(sa_target(), sa_data()) -> {any(), sa_target(), sa_data()}.
+next(Target, Data) ->
+  Temperature = Data#sa_data.temperature,
+  NextGenerator = (Target#sa_target.next)(Target#sa_target.last_generated,
+                                          Temperature),
```

Comment:
The four lines above are a bit ugly.  I suggest the following rewrite:
```erlang
next(#sa_target{next = Next, last_generated = LastGen} = Target, Data) ->
  NextGenerator = Next(LastGen, Data#sa_data.temperature),
```
which is both prettier and half the size (in lines).